pandas/src/hashtable_class_helper.pxi.in

```diff
@@ -864,3 +836,109 @@ cdef class PyObjectHashTable(HashTable):
                 count += 1
 
         return np.asarray(labels)
+
+
+cdef class MultiIndexHashTable(HashTable):
+
+    def __init__(self, size_hint=1):
+        self.table = kh_init_uint64()
+        self.mi = None
+        kh_resize_uint64(self.table, size_hint)
+
+    def __dealloc__(self):
+        if self.table is not NULL:
+            kh_destroy_uint64(self.table)
+            self.table = NULL
+
+    def __len__(self):
+        return self.table.size
+
+    def sizeof(self, deep=False):
+        """ return the size of my table in bytes """
+        return self.table.n_buckets * (sizeof(uint64_t) + # keys
+                                       sizeof(size_t) + # vals
+                                       sizeof(uint32_t)) # flags
+
+    def __contains__(self, object key):
+        cdef:
+            khiter_t k
+            uint64_t value
+
+        from pandas.tools.hashing import hash_tuples
+        key = self.mi._as_valid_indexing_key(key)
+        value = hash_tuples(key)
+        k = kh_get_uint64(self.table, key)
+        return k != self.table.n_buckets
+
+    cpdef get_item(self, object key):
+        cdef:
+            khiter_t k
+            uint64_t value
+
+        from pandas.tools.hashing import hash_tuples
+        key = self.mi._as_valid_indexing_key(key)
+        value = hash_tuples(key)
+
+        # should check nlevels == len tuple
+        k = kh_get_uint64(self.table, value)
```

Comment:
it will slow down worst case checking a bit (still about 2x faster than master)

```
In [1]: i = MultiIndex.from_product([np.arange(1000), np.arange(1000)], names=['one','two'])

In [2]: v = i.values

In [3]: %time i.get_indexer(v)
CPU times: user 356 ms, sys: 84.4 ms, total: 441 ms
Wall time: 441 ms
Out[3]: array([     0,      1,      2, ..., 999997, 999998, 999999])
```

but type of hit would *only* occur if you are actually reindexing from tuples themselves (and not an actual MI).